libr/parse/p/parse_riscv_pseudo.c

```diff
@@ -0,0 +1,229 @@
+/* radare - LGPL - Copyright 2020 - Aswin C (officialcjunior) */
+
+#include <r_lib.h>
+#include <r_flag.h>
+#include <r_anal.h>
+#include <r_parse.h>
+
+static int replace(int argc, const char *argv[], char *newstr) {
+#define MAXPSEUDOOPS 10
+	int i, j, k, d;
+	char ch;
+	struct {
+		int narg;
+		char *op;
+		char *str;
+		int args[MAXPSEUDOOPS];
+	} ops[] = {
+		{ 0, "add", "# = # + #", { 1, 2, 3 } },
+		{ 0, "addi", "# = # + #", { 1, 2, 3 } },
+		{ 0, "and", "# = # & #", { 1, 2, 3 } },
+		{ 0, "andi", "# = # & #", { 1, 2, 3 } },
+		{ 0, "beq", "if (# == #) goto #", { 1, 2, 3 } },
+		{ 0, "bleu", "if (unsigned)# <= # goto #", { 1, 2, 3 } },
+		{ 0, "bltu", "if (unsigned)# < # goto #", { 1, 2, 3 } },
+		{ 0, "blt", "if (# < #) goto #", { 1, 2, 3 } },
+		{ 0, "beqz", "if (# == 0) goto #", { 1, 2 } },
+		{ 0, "bne", "if (# != #) goto #", { 1, 2, 3 } },
+		{ 0, "bnez", "if (# != 0) goto #", { 1, 2 } },
+		{ 0, "bgez", "if (# >= 0) goto #", { 1, 2 } },
+		{ 0, "bgtz", "if (# > 0) goto #", { 1, 2 } },
+		{ 0, "fld", "# = #", { 1, 2 } },
+		{ 0, "j", "jmp #", { 1 } },
+		{ 0, "jr", "jmp #", { 1 } },
+		{ 0, "jalr", "jmp #", { 1 } },
+		{ 0, "jal", "jmp #", { 1 } },
+		{ 0, "ld", "# = (double)[#]", { 1, 2 } },
+		{ 0, "li", "# = #", { 1, 2 } },
+		{ 0, "lh", "# = [#]", { 1, 2 } },
+		{ 0, "lui", "# = #", { 1, 2 } },
+		{ 0, "lbu", "# = (unsigned)[#]", { 1, 2 } },
+		{ 0, "lhu", "# = (unsigned)[#]", { 1, 2 } },
+		{ 0, "lw", "# = [#]", { 1, 2 } },
+		{ 0, "mv", "# = #", { 1, 2 } },
+		{ 0, "or", "# = # | #", { 1, 2, 3 } },
+		{ 0, "sd", "[#] = (double)#", { 2, 1 } },
+		{ 0, "sw", "[#] = #", { 2, 1 } },
+		{ 0, "sb", "[#] = #", { 2, 1 } },
+		{ 0, "sh", "[#] = #", { 2, 1 } },
+		{ 0, "sub", "# = # - #", { 1, 2, 3 } },
+		{ 0, NULL }
+	};
+	if (!newstr) {
+		return false;
+	}
+
+	for (i = 0; ops[i].op; i++) {
+		if (ops[i].narg) {
+			if (argc - 1 != ops[i].narg) {
+				continue;
+			}
+		}
+		if (!strcmp (ops[i].op, argv[0])) {
+			if (newstr) {
+				d = 0;
+				j = 0;
+				ch = ops[i].str[j];
+				for (j = 0, k = 0; ch != '\0'; j++, k++) {
+					ch = ops[i].str[j];
```

Comment:
Too many levels of nesting - please split to a smaller blocks, functions, etc.

Reply:
Thank you for your reply. I really appreciate it.

So, I was looking up about functions similar to what you are referring to and saw `can_replace()` in `parse_mips_pseudo` which I believe checks if something can be replaced. Is that what you're talking about here? Or is it the `if (ch == '#')` part?

If not, can you just write what and which exact part should be done by a function?

Reply:
he is refering to moving the body of the function to a separate function. so the whole function fits a single screen

Reply:
but as i said in my previous comment, i dont think refactoring this is really necessary because this code is duplicated multiple times and this interesting and useful feature may take ages to be merged if this cleanup ends in the full refactoring that implies having a common pseudo parser for all archs so we dont duplicate code

Reply:
Marking as resolved.